lib/client/conntest/kube.go

```diff
@@ -0,0 +1,273 @@
+			// For the happy path, we do update the Message to Success.
+			Message: types.DiagnosticMessageFailed,
+		})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := s.cfg.UserClient.CreateConnectionDiagnostic(ctx, connectionDiagnostic); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	currentUser, err := s.cfg.UserClient.GetCurrentUser(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	key, err := client.GenerateRSAKey()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	certs, err := s.cfg.UserClient.GenerateUserCerts(ctx, proto.UserCertsRequest{
+		PublicKey:              key.MarshalSSHPublicKey(),
+		Username:               currentUser.GetName(),
+		Expires:                time.Now().Add(time.Minute).UTC(),
+		ConnectionDiagnosticID: connectionDiagnosticID,
+		KubernetesCluster:      req.ResourceName,
+	})
+
+	diag, diagErr := s.handleUserGenCertsErr(ctx, connectionDiagnosticID, err)
+	if err != nil || diagErr != nil {
+		return diag, diagErr
+	}
+
+	key.TLSCert = certs.TLS
+
+	certAuths, err := s.cfg.UserClient.GetCertAuthorities(ctx, types.HostCA, false /* loadKeys */)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	key.TrustedCA = auth.AuthoritiesToTrustedCerts(certAuths)
+	ca, err := s.cfg.UserClient.GetClusterCACert(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	ctxWithTimeout, cancelFunc := context.WithTimeout(ctx, req.DialTimeout)
+	defer cancelFunc()
+
+	restConfig := &rest.Config{
+		Host: "https://" + s.cfg.KubernetesPublicProxyHostPort,
+		TLSClientConfig: rest.TLSClientConfig{
+			CAData:   ca.TLSCA,
+			CertData: key.TLSCert,
+			KeyData:  key.PrivateKeyPEM(),
+		},
+	}
+
+	if s.cfg.TLSRoutingEnabled {
+		k8host := strings.Split(s.webProxyAddr, ":")[0]
+		// replace localhost with "127.0.0.1" so GetKubeTLSServerName can generate a domain kube.cluster.local.
+		if strings.EqualFold(k8host, "localhost") {
+			k8host = "127.0.0.1"
+		}
+		restConfig.TLSClientConfig.ServerName = client.GetKubeTLSServerName(k8host)
+		restConfig.Host = "https://" + s.webProxyAddr
+	}
+
+	client, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	_, err = client.CoreV1().Pods(req.KubernetesNamespace).List(ctxWithTimeout, v1.ListOptions{})
+	diag, diagErr = s.handleErrFromKube(ctxWithTimeout, connectionDiagnosticID, err, req.KubernetesNamespace)
+	if err != nil || diagErr != nil {
+		return diag, diagErr
+	}
+
+	traceType := types.ConnectionDiagnosticTrace_KUBE_PRINCIPAL
+	message := "Access to the Kubernetes Cluster granted."
+	connDiag, err := s.appendDiagnosticTrace(ctxWithTimeout, connectionDiagnosticID, traceType, message, nil)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	connDiag.SetMessage(types.DiagnosticMessageSuccess)
+	connDiag.SetSuccess(true)
+
+	if err := s.cfg.UserClient.UpdateConnectionDiagnostic(ctxWithTimeout, connDiag); err != nil {
+		return nil, trace.Wrap(err)
```

Comment:
This is related to `SSH` connection testing mechanism. 

Since the SSH conn testing offloads the execution to `tsh ssh ....` we are required to init the testing and embed the `connectionDiagnosticID` inside the SSH Certificate extension. Teleport SSH service reads the `connectionDiagnosticID` from the certificate and appends the desired data and steps into the disk so we can extract it at the ssh connection testing once it finishes.

I followed the same global idea although here we create a K8S client directly to kubeproxy.

Reply:
I see, in that case I agree that consistent behavior between the two diagnostics is best.